_posts/2019/2019-05-11-how-to-improve-your-components-with-react-hooks.md

```diff
@@ -0,0 +1,183 @@
+    </div>
+  );
+}
+```
+
+We can immediately notice that our component got shorter, which is usually a good thing. But I would also argue that our component got easier to read and understand. Of course, this is always going to be subjective. But for the sake of the argument, let me hypothesize why one might get this feeling of reduced complexity.
+
+I think the reason that this function version of our component is easier to understand lies in **the way the component logic is organized**. To demonstrate this point, I put both versions side by side and added some color coding. (Yellow: Code related to user data, Red: Code related to screen size)
+
+![Component comparison](/assets/img/pages/blog/images/react-hooks-component-comparison.png)
+
+We can see that in the function component, each logic is much more grouped together than in the class component. This not only makes our component easier to read top to bottom but also easier to extract logic from.
+
+#### Going further
+
+Now that we have rewritten our component as a function, let's think about how we can improve it even further.
+
+Let's say we wanted to reuse some of our component logic in another place. We could decide to just copy and paste the logic we want to reuse, but React now provides us with a better way: Custom Hooks. Custom Hooks allow you to extract your component logic into reusable functions. Here's how they would look for our `UserProfile` component:
+
+```js
+function useUser(userId) {
+  const [user, setUser] = React.useState(null);
+
+  React.useEffect(() => {
+    fetch("https://some-api.com/user/" + userId)
+      .then(response => response.json())
+      .then(user => setUser(user));
+  }, [userId]);
+
+  return user;
+}
+
+function useResponsive() {
+  const [isMobile, setIsMobile] = React.useState(window.innerWidth <= 1024);
+
+  React.useEffect(() => {
+    const handleResize = () => {
+      if (isMobile && window.innerWidth > 1024) {
+        setIsMobile(false);
+      } else if (!isMobile && window.innerWidth <= 1024) {
+        setIsMobile(true);
+      }
+    };
+    window.addEventListener("resize", handleResize);
+    return () => window.removeEventListener("resize", handleResize);
+  }, []);
+
+  return isMobile;
+}
+```
+
+Now that we have extracted the different pieces of component logic, we can reuse them in our `UserProfile` component.
+
+```js
+function UserProfile({ userId }) {
+  const user = useUser(userId);
+  const isMobile = useResponsive();
+
+  if (!user) return null;
+  return (
+    <div>
+      {!isMobile && <img src={user.image} />}
+      <p>{user.name}</p>
+    </div>
+  );
+}
+```
+
+On top of improving the readability of our `UserProfile` even further, we can now reuse our logic in any component we want to use it. 🎉
+
+<br />
+
+**Even though we haven't been rewriting all our apps with Hooks here at ePages, we have already found great value in using them for our refactors and new components. We are hooked, are you?**
```

Comment:
```suggestion
**Even though we haven't been rewriting all our apps with Hooks here at ePages, we have already found great value in using them for our refactors and new components. We are hooked, what about you?**
```

Reply:
Do you think this sounds better or is this actually a grammatical mistake by me?

Reply:
I just think it sounds better.